### PR TITLE
docs: Rest api docs.

### PR DESCRIPTION
**Pull request by chriscm2006**
## Requester Review Items

- [x] Ensure the Pull Request is into develop.
- [x] Test coverage has not gone down.
- [x] Documentation updated or not needed. (Wiki, Issues, etc)
- [x] Angular Type leads to proper Semantic Version.

## Dev Team Review Items

To be filled out by @dequelabs/mobile.

- [ ] Ensure the Requester did not lie. Chastise them politely if they did.
- [ ] Code Quality review.
- [ ] Changes to Axe*.java classes are minimal, appropriate, and versioned properly.
  - [ ] Serialized Axe objects have not changed.
  - [ ] Public API has not changed.

## Code Owner Review Items

To be filled out by @dequelabs/mobileadmin.

- [ ] Any Rules package changes lead to proper Semantic Version.
- [ ] Security Review.

## Merger Review Items

- [ ] Ensure commit message matches title before squashing.

### CODE REVIEW

**Comment by jeankaplansky** [comment]
DPI is very much print terminology. Is it helpful to you guys to think in terms of PPI (pixels per inch)? Or is that still going to be messed up because of Retina/HD screens and the algorithms they do to make their bazillion PPIs LOOK like 44 PPI?

**Comment by neocln** [comment]
Overall looks reasonable. One question is why an event and a hierarchy are contained in a single result. to my view, a correlation between them is weak. Maybe it is due to my lack of Android AT APIs. in Windows world, there is a set of APIs to get a specific UI element at a moment based on focus or UI element ID(Automation ID/Runtime ID) or full hierarchy tree from the top(desktop). because of that when a scan is invoked, there is no reason to have a correlation between a scan result and an event. I wonder whether a similar thing can be done on Android.

File: docs/Rest-API.md [hunk 1]
```diff
@@ -0,0 +1,147 @@
...
+
+- All motivations of AxeContext apply.
+- Rule debugging.
+- Simplified test case collection.
+- Simplified false positive identification.
+
+### Simplified Sample
+
+```
+{
```
Comment (karanbirsingh) [inline docs/Rest-API.md]: How will elements be identified in AXRL? It seems like a path-based approach would be possible with this view representation; what are thoughts about an ID for each node in the tree?

File: docs/Rest-API.md [hunk 2]
```diff
@@ -0,0 +1,147 @@
...
+      "children" : []
+    }
+  ]
+}
+```
+
+### Requirements
+
+- The Arbitrary property names of the children are identical to the root. 
+- If a property is omitted from a child it is assumed to be present and NULL.
```
Comment (RobGallo) [inline docs/Rest-API.md]:
```suggestion
- If a property is omitted from a child it is assumed to be not present or NULL.
```

File: docs/Rest-API.md [hunk 3]
```diff
@@ -0,0 +1,147 @@
...
+
+### Motivation
+
+- Identify device version and operating system.
+- Provide metrics for screen size so view hierarchy can be correlated.
+
+### Simplified Sample
+
+```
+{
```
Comment (RobGallo) [inline docs/Rest-API.md]: Should/can we include the device model? Sometimes the same OS can behave differently on different devices.

File: docs/Rest-API.md [hunk 4]
```diff
@@ -0,0 +1,147 @@
...
+
+- Collect View information in one place.
+- Allow for direct correlation of screenshot and ViewHierarchy
+- Easy developer debugging.
+- Reproducability of results.
+
+### Simplified Sample
+
+```
+{
```
Comment (RobGallo) [inline docs/Rest-API.md]: Should we include any meta data about the scan itself, e.g., version of Axe, duration, errors? This information can be useful for posterity.

File: docs/Rest-API.md [hunk 5]
```diff
@@ -0,0 +1,147 @@
...
+  }
+}
+
+```
+
+### Requirements
+
+- Property names match the above exactly.
+
+## AxeEvents
```
Comment (RobGallo) [inline docs/Rest-API.md]: If this section is intended to imply multiple scans per session, timing information, or at least some type of corelation between scan results and events will be necessary. Is this also intended to imply that scans may be run based on events?

File: docs/Rest-API.md [hunk 6]
```diff
@@ -0,0 +1,147 @@
+# Axe REST API
```
Comment (RobGallo) [inline docs/Rest-API.md]: Overall, I think this document needs a lot more context. For example, a rest endpoint between what and what? What is the overall purpose of the spec? How about examples of what "arbitrary value" could be? What is the collection of results intended to cover: a scan of an entire app, a single element, multiple scans of individual elements,an entire app, multiple apps? How does a certain design make things easier for developers?

File: docs/Rest-API.md [hunk 7]
```diff
@@ -0,0 +1,147 @@
...
+### Motivation
+
+- Provide access to results and context easily.
+- Avoid timing issues that come with multiple requests.
+
+### Simplified Sample
+```
+{
+  "axeContext" : { @AxeContext },
+  "axeResult" : { AXRL },
```
Comment (RobGallo) [inline docs/Rest-API.md]: Does AxeResult include information on what app was scanned? I didn't find that anywhere in this doc. How about a link to the AXRL doc?

Comment (chriscm2006) [reply docs/Rest-API.md]: AxeContext includes this information. A link to the AXRL doc is a good idea. I do not actually plan on hosting this here, I think it belongs elsewhere, and so this doc is inherently mildly incomplete.

**Comment by chriscm2006** [comment]
> Overall looks reasonable. One question is why an event and a hierarchy are contained in a single result. to my view, a correlation between them is weak. Maybe it is due to my lack of Android AT APIs. in Windows world, there is a set of APIs to get a specific UI element at a moment based on focus or UI element ID(Automation ID/Runtime ID) or full hierarchy tree from the top(desktop). because of that when a scan is invoked, there is no reason to have a correlation between a scan result and an event. I wonder whether a similar thing can be done on Android.

@neocln You're correct. The correlation is weak, but it is still valuable. For the sake of `Event Based` rules, they are based on the point at which a "Window Update" event has occured. At that point we clear the event stream and start recording events that happen. As such, for the sake of `Event Based` rule calculation the events can be thought of belonging grouped on a timeline between the last `Window Content Update Android Event` and a `User Requested Analysis`.

This corelation is weak, however, without some type of timestamp or correlation, cmomunicating this information becomes meaningless for the sake of running Axe rules against this hierarchy. 

We could provide independent REST endpoints for accessing all of this information in an uncorrelated way, though I'm not sure that provides additional value in an Axe context. I'd have to understand the use case.

File: docs/Rest-API.md (hunk 4 above)
Comment (chriscm2006) [reply docs/Rest-API.md]: Hmm. My question is where does this belong. At this level, or one level up. Good idea though. We will add this.

File: docs/Rest-API.md (hunk 3 above)
Comment (chriscm2006) [reply docs/Rest-API.md]: This REST API is purposely designed to be very flexible. Particularly between iOS and Android. Any time a property is listed specifically it is required to be exactly that. Adding "model" to this would potentially create issues for iOS. 

What would you think about going with a non-structured "description" approach, which would include this information in a non-structured way for compatibility and clarity between the two frameworks.

File: docs/Rest-API.md (hunk 5 above)
Comment (chriscm2006) [reply docs/Rest-API.md]: No, events are intended to mean `Accessibility Events` for **Android** and `Accessibility Notifications` for **iOS**.

File: docs/Rest-API.md (hunk 6 above)
Comment (chriscm2006) [reply docs/Rest-API.md]: If there's Context you feel your are lacking, given the fact that this doesn't exist yet, a quick call would be great! :)

File: docs/Rest-API.md (hunk 1 above)
Comment (RobGallo) [reply docs/Rest-API.md]: The DRY principle does not apply to documentation. Karan's question is valuable, and indicates information which is not readily apparent and which, at the very least, should be referenced.

File: docs/Rest-API.md [hunk 8]
```diff
@@ -0,0 +1,147 @@
+# Axe REST API
+
+Documentation of the REST API that comes along with the Axe service.
```
Comment (DaveTryon) [inline docs/Rest-API.md]: >REST API [](start = 21, length = 8)

What REST verb is being documented here? It feels like GET, but the docs don't say. Also, what parameters, if any, should be specified by the request? For example, is the scan always the entire display or just part of it? What about multi-screen devices where different screens could potentially have different apps/views? The Android documentation says that Android doesn't support a "root" or "desktop" view, so would multiple apps require a collection of separate hierarchies?

File: docs/Rest-API.md [hunk 9]
```diff
@@ -0,0 +1,147 @@
...
+
+## AxeResult
+
+This endpoint serves as a wrapper to collect all of the information from an axe run.
```
Comment (DaveTryon) [inline docs/Rest-API.md]: >endpoint [](start = 5, length = 8)

The docs say "endpoint", then appear to describe the response of the REST API. Shouldn't this document include the information necessary to actually submit the request? Consider, for example, the GitHub API documentation, which include both sample requests and sample responses: [https://developer.github.com/v3/](https://developer.github.com/v3/)

File: docs/Rest-API.md [hunk 10]
```diff
@@ -0,0 +1,147 @@
...
+Documentation of the REST API that comes along with the Axe service.
+
+## AxeResult
+
+This endpoint serves as a wrapper to collect all of the information from an axe run. 
+
+### Motivation
+
+- Provide access to results and context easily.
+- Avoid timing issues that come with multiple requests.
```
Comment (DaveTryon) [inline docs/Rest-API.md]: This is the only and only reference to timing in the entire document, and the comment seems incomplete as a result without something to at least provide a one-sentence summary of how the timing issues are handled. For example, what happens if I have a multi-threaded app that makes concurrent calls to this API? If one fails, what form will that failure take?

File: docs/Rest-API.md (hunk 7 above)
Comment (DaveTryon) [reply docs/Rest-API.md]: Absolutely agree with Rob that this document is significantly incomplete without more information about the AXRL format

---
In reply to: [274424144](https://github.com/dequelabs/axe-android/pull/11#discussion_r274424144) [](ancestors = 274424144)

File: docs/Rest-API.md [hunk 11]
```diff
@@ -0,0 +1,147 @@
...
+
+### Simplified Sample
+
+```
+{
+  "Arbitrary Property Name": "value1",
+  "Arbitrary Property Name2": "value2",
+  "children" : [
+    {
+      "Arbitrary Property Name": "value3",
```
Comment (DaveTryon) [inline docs/Rest-API.md]: >"Arbitrary Property Name": "value3", [](start = 6, length = 36)

A comment more about the design than the document: A string to string mapping makes for excellent portability for the data generator, but horrible portability for the data consumer. For example, it's likely that Android, iOS, and Windows will have multiple overlapping concepts as properties, but with potentially different property names and/or data formats. In order to create a single viewer that can support all of the variants, we'll probably need to convert the platform-specific data into a platform-agnostic format to keep the view code sane

File: docs/Rest-API.md (hunk 3 above)
Comment (DaveTryon) [reply docs/Rest-API.md]: Trying to parse meaningful data from non-structured data is something that I'd strongly suggest that we avoid because it's inherently brittle. In my experience, you can safely do only 2 things with non-structured fields:
1. Display them in the UI
2. Group by them in reports/telemetry

If we want the client to offer greater functionality (such as programmatically identifying this as a Samsung devices), then a dedicated field--even an optional one--is far more robust.

---
In reply to: [279103399](https://github.com/dequelabs/axe-android/pull/11#discussion_r279103399) [](ancestors = 279103399)

File: docs/Rest-API.md [hunk 12]
```diff
@@ -0,0 +1,147 @@
...
+- Identify device version and operating system.
+- Provide metrics for screen size so view hierarchy can be correlated.
+
+### Simplified Sample
+
+```
+{
+  "os": {
+    "name" : "Android",
+    "version" : "8.1.0"
```
Comment (DaveTryon) [inline docs/Rest-API.md]: >8.1.0 [](start = 17, length = 5)

Is this field structured? If so, please define the structure. If not, see my previous comments about the limited usefulness of unstructured data.

File: docs/Rest-API.md [hunk 13]
```diff
@@ -0,0 +1,147 @@
...
+  },
+  "screen" : {
```
Comment (DaveTryon) [inline docs/Rest-API.md]: >screen [](start = 3, length = 6)

How will multi-screen devices be represented?

File: docs/Rest-API.md [hunk 14]
```diff
@@ -0,0 +1,147 @@
...
+    "width" : 1080,
+    "dpi": 2.625
+  }
+}
+
+```
+
+### Requirements
+
+- Property names match the above exactly.
```
Comment (DaveTryon) [inline docs/Rest-API.md]: >Property names [](start = 2, length = 14)

Are property names case sensitive?

File: docs/Rest-API.md [hunk 15]
```diff
@@ -0,0 +1,147 @@
...
+
+## AxeEvents
+
+Along with a view hierarchy modern Operating Systems communicate with Assistive Technologies through
+events. Should contain events since the last Screen Changed event.
```
Comment (DaveTryon) [inline docs/Rest-API.md]: >Screen Changed [](start = 45, length = 14)

The calls to the API and the Screen Changed events appear to be completely unrelated streams. I'm not seeing the value of this feature unless there's at least a way to join the two streams chronologically.

File: docs/Rest-API.md [hunk 16]
```diff
@@ -0,0 +1,147 @@
...
+
+### Motivation
+
+- Similar motivations to AxeView, but applied to the event stream.
```
Comment (DaveTryon) [inline docs/Rest-API.md]: >stream [](start = 59, length = 6)

Hopefully there will be a way to filter the events that get inserted into the stream? On Windows, for example, we only register event handlers when requested, for performance reasons.

File: docs/Rest-API.md [hunk 17]
```diff
@@ -0,0 +1,147 @@
...
+- Collect dynamic information about the view and how its evolving.
+
+### Simplified Sample
```
Comment (DaveTryon) [inline docs/Rest-API.md]: >Sample [](start = 15, length = 6)

Some things (like timing and perhaps the hash of the item associated with the event) seem pretty universal here, and I wonder if they ought to be promoted to first-class citizens, instead of being an optional item in a property bag.

File: docs/Rest-API.md [hunk 18]
```diff
@@ -0,0 +1,147 @@
...
+### Motivation
+
+- Identify device version and operating system.
+- Provide metrics for screen size so view hierarchy can be correlated.
+
+### Simplified Sample
+
+```
+{
+  "os": {
```
Comment (DaveTryon) [inline docs/Rest-API.md]: >os [](start = 3, length = 2)

You might want to include a property bag for platform-specific stuff that may neatly fit into the fields that you're defining here.

File: docs/Rest-API.md (hunk 1 above)
Comment (chriscm2006) [reply docs/Rest-API.md]: You're lacking context in which this document will actually be published. I understand your concern, if it's still a concern when this gets published we will address it then, for now, we should drop this conversation.

**Comment by chriscm2006** [comment]
I'm gonna close this. @DaveTryon is correct. It is more of a spec. Tho docs will be published within a greater documentation site. Without the context for that most of this coversation is unproductive, and I don't want to publish this in two places. 

The spec input was valuable and will be taken into consideration for the final version. Thanks all!
